### PR TITLE
Add GitHub Actions documentation-deployment pipeline

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This brings a documentation-deployment pipeline into the Qiskit/Terra repository, allowing it to fully deploy its documentation to `qiskit.org`, a task previously only the metapackage could perform. This does not fully unify the documentation with files from the metapackage, it just adds a pipeline to do the final deployment.

This includes a revitalised translatable-strings pipeline, which was previously broken on the metapackage for the last month or two. It also previously included a fair amount of legacy weight that was no longer relevant.

### Details and comments

- Fix Qiskit/qiskit-metapackage#1779
- Fix Qiskit/qiskit-metapackage#1777
- Fix Qiskit/qiskit-metapackage#1746 (since they become auto hooked up)
- Fix Qiskit/qiskit-metapackage#1724
- Fix Qiskit/qiskit-metapackage#1717 (via the new deployment logic doing the deployments to `/` and `/stable/0.x` together)

... although spiritually they're not _totally_ fixed until we also do Qiskit/qiskit-metapackage#1722, which is most likely coming after this PR, since the deployment of the documentation would still be missing a bunch of top-level files (like the landing page) right now.

This needs the secrets for decrypting the rclone remote config and the SSH key with write access on qiskit-community/qiskit-translations, although in the new form, it may be cleaner to add the SSH key as a direct secret on Terra, since then we can avoid the manual step to mark it as secret (masked) text when moving the decryption into the git credentials.

You can see an example run here: https://github.com/jakelishman/qiskit-terra/actions/runs/5825390347.  (Obviously the actual pushes fail due to missing secrets / permissions.)

*edit*: **The following is out-of-date and has been superseded by https://github.com/Qiskit/qiskit-terra/pull/10610#issuecomment-1675085149.**

The strategy for pushing docs to `qiskit.org/documentation` here is:
- on a `push` event to `main`, deploy to `/dev`
- on a `push` event to _any_ `stable/*`, deploy to `/` - this needs revisiting when we start supporting more than one branch
- on a `tag` event, deploy to the relevant `/stable/<major>.<minor>` - can be changed if we decide to only push a single form for each major
- if run manually, the documentation prefix can be set to anything to push to `qiskit.org/documentation/<prefix>`, though in practice anything other than `/`, `/stable` or `/dev` would get removed on the next push to `/`.

We push the translatable strings on any push to `stable/*`, which has the same caveat as the docs push above.

### CODE REVIEW

File: tools/docs_exclude.txt
Comment (jakelishman) [inline tools/docs_exclude.txt:3]: Note that this file is completely different to the one currently on the metapackage, since we're no longer pushing various applications modules / elements to the root of `qiskit.org/documentation` any more.  A deployment with this drastically stripped-down `docs_exclude.txt` will cause some things (like https://qiskit.org/documentation/retworkx) that have currently just about survived to be deleted.

File: .github/workflows/docs_deploy.yml
Comment (mtreinish) [inline .github/workflows/docs_deploy.yml:123]: So any push to `stable/*` will update the main published version of the docs? I'm a bit concerned that for somethings we might be changing the docs and undoing it for things like release notes as they'll be auto-generated via reno in the near future. Also for release candidates we're working on stable before we've released the version so the docs on the main page will reflect this. I'm thinking it might be better to reserve this for only on tag events and workflow dispatch.

File: tools/docs_exclude.txt
Comment (mtreinish) [reply tools/docs_exclude.txt:3]: We'll need to wait until we have all the redirects in place before deploy then. Right now there are some pages on `documentation/*` that have survived as you point out and we don't want links to those pages to 404.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, the deployment scripts seem fine. My only real concern is about the deployment strategy on stable branches.

File: tools/docs_exclude.txt
Comment (jakelishman) [reply tools/docs_exclude.txt:3]: Luciano's on top of this, I believe.  Fwiw, the only things affected are `retworkx`, `metal`, `neko` and the index page of `partners`.

We also shouldn't add the secrets or actually enable deployment from this PR until the docs build is actually fully complete on Terra, which is still at least a PR away.

Comment (mtreinish) [reply tools/docs_exclude.txt:3]: The secrets are actually already populated. I added them 3 weeks ago in anticipation of this.

File: .github/workflows/docs_deploy.yml
Comment (jakelishman) [reply .github/workflows/docs_deploy.yml:123]: Oh yeah, I didn't think about the release-candidate phase.

I'll have to rewrite the logic a bit here to allow more than one deployment to happen from one event; we'll need `tag` to trigger two different uploads with your strategy (which is fine, I just need to rewrite the logic).  Previously the metapackage had similar-ish logic to what I've written, in that any push to `master` triggered a deployment to `/`, so there wasn't actually ever a double deployment from one action trigger, it's just that a push and a tag both triggered deployments.

File: tools/docs_exclude.txt
Comment (jakelishman) [reply tools/docs_exclude.txt:3]: I think you'll need to check the translatable-strings ones have the same names I gave them - I think I might have renamed them by accident.

Comment (jakelishman) [reply tools/docs_exclude.txt:3]: But if you've added the secrets, then we definitely shouldn't allow (at the very least) the push events from this workflow to fire, and most likely we should only merge this PR after the documentation is actually ready to deploy.

tbh, that was probably the more sensible ordering from the start, I just thought this PR would be way easier to write than it was.

File: .github/workflows/docs_deploy.yml
Comment (jakelishman) [reply .github/workflows/docs_deploy.yml]: Done in fc27c4a8f.

**Comment by jakelishman** [comment]
Ok, fc27c4a8f improves the logic for deployments significantly and fixes a bug that would have caused tag-push events to fail, because GitHub calls receiving a tag event a "push" (totally validly), but I'd accidentally referred to it as "tag".

The new logic is:
- push to `main` triggers docs deployment to `/dev`
- push to `stable/*` does not trigger anything
- a full-release tag (`[0-9]+\.[0-9]+\.[0-9]+`) triggers a build, and deploys docs to the relevant `stable/*` prefix.  If and only if the tag is the _latest_ version known to git, disregarding prereleases of newer versions, the docs are also deployed to `/` and the translatable strings are pushed
- prerelease tags do not trigger anything
- a manual run can be from anywhere, and has tick boxes to decide whether to push to qiskit.org and/or the translatable strings.

Example runs, on a repo whose highest version tag before any of these triggers was `0.17.4`, but also contained a prerelease tag `0.18.0rc1`:
- [push to `main` deploys docs to `/dev`](https://github.com/jakelishman/qiskit-terra/actions/runs/5834846017)
- [push to newest non-prerelease tag (`0.17.5`) triggers docs deployments to `/stable/0.17` and `/`, and translatable strings push](https://github.com/jakelishman/qiskit-terra/actions/runs/5834846267)
- [push to not the newest tag (`0.17.4`) triggers docs deployment to `/stable/0.17` only](https://github.com/jakelishman/qiskit-terra/actions/runs/5834846220) (note that I should have demonstrated this using a `0.16.5` tag or something - we'd never release a tag that was already outdated within its patch series)
- [manual trigger to a deployment prefix "", requesting both docs deployment and translations](https://github.com/jakelishman/qiskit-terra/actions/runs/5834855624)

In all those runs, I cut down the actual builds to not take forever, and I cut out all the logic that would _actually_ push and just turned it into echoes, etc.

**Comment by jakelishman** [comment]
This is ready to review, I've just marked it on hold because the secrets are in place on Terra, but we're not actually ready for the auto-deployment to start triggering yet.  There's still some old documentation redirects that need configuring on qiskit.org, and we need to bring in the last components of the full documentation build from the metapackage - at the moment, it'd deploy an ugly old version of the docs without the correct index pages.

**Comment by jakelishman** [comment]
The new deployment logic causes this PR to fix Qiskit/qiskit-metapackage#1717.

**Comment by jakelishman** [comment]
Ok, merge conflicts with #10611 are now resolved and that is merged, so this repository now builds a complete version of the documentation.  With the new deployment logic meaning that only a new tag event or manual deployment to `/` would wipe out any existing documentation sites (plus the reconfiguration of `qiskit.org` to add the new redirects should merge soon), I think we can remove "on hold" and this is ready to go.

File: .github/workflows/docs_deploy.yml
Comment (mtreinish) [inline .github/workflows/docs_deploy.yml:8]: If it was a regex I'd want to end this in `$` to assert there is no `rc` or other post notation. I looked at the docs for this: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet and it's not a regex and this should work. This also precludes more than 10 patch versions,  but I don't think we've ever needed that. If we do, we can revisit the docs publishing behavior at that time.

File: .github/workflows/docs_deploy.yml
Comment (mtreinish) [inline .github/workflows/docs_deploy.yml:131]: I initially did a double take on this thinking it would cause failures on `stable/*` branches. But then I remembered the job is only set to trigger on pushes to `main`.

**Review by mtreinish — APPROVED** [review]
This LGTM, I had a couple of inline comments but more just my confirmation of how things worked rather than things to change.

**Pull request merged**
